File: Diagrammatize.Design.Pattern/code/cxx/Proxy/Sample/PrinterProxy.cpp
#include "PrinterProxy.hpp"

PrinterProxy::PrinterProxy() {
}

PrinterProxy::PrinterProxy(const std::string& name_): name(name_) {
}

void PrinterProxy::setPrinterName(const std::string& name_) {
    std::lock_guard<std::mutex> guard(lock);
    if (real != nullptr) {
        real->setPrinterName(name_);
    }
    name = name_;
}

std::string PrinterProxy::getPrinterName() {
    return name;
}

void PrinterProxy::print(const std::string& string) {
    realize();
    real->print(string);
}

void PrinterProxy::realize() {
    std::lock_guard<std::mutex> guard(lock);
    if (real == nullptr) {
        real = new Printer(name);
    }
}

File: Diagrammatize.Design.Pattern/code/python/Proxy/Sample/PrinterProxy.py
import threading

from Printable import Printable
from Printer import Printer

class PrinterProxy(Printable):
    def __init__(self, name=""):
        self.name = name
        self.real = None
        self.lock = threading.Lock()

    def setPrinterName(self, name):
        with self.lock:
            if self.real is not None:
                self.real.setPrinterName(name)
            self.name = name

    def getPrinterName(self):
        return self.name

    def print(self, string):
        self.realize()
        self.real.print(string)

    def realize(self):
        with self.lock:
            if self.real is None:
                self.real = Printer(self.name)
